consensus/safety-rules/src/lib.rs

```diff
@@ -201,6 +165,44 @@ impl SafetyRules {
         None
     }
 
+    /// Produces a LedgerInfo that either commits a block based upon commit_rule_for_certified_block
+    /// or an empty LedgerInfo for no commit
+    fn construct_ledger_info<T: Payload>(&self, proposed_block: &Block<T>) -> LedgerInfo {
+        let potential_commit_id = self
+            .commit_rule_for_certified_block(proposed_block.quorum_cert(), proposed_block.round());
+
+        match potential_commit_id {
+            Some(_potential_commit_id) => LedgerInfo::new(
+                proposed_block.quorum_cert().parent_block().version(),
+                proposed_block
+                    .quorum_cert()
+                    .parent_block()
+                    .executed_state_id(),
+                HashValue::zero(),
+                proposed_block.quorum_cert().parent_block().id(),
+                proposed_block.quorum_cert().parent_block().epoch(),
+                proposed_block
+                    .quorum_cert()
+                    .parent_block()
+                    .timestamp_usecs(),
+                proposed_block
+                    .quorum_cert()
+                    .parent_block()
+                    .next_validator_set()
+                    .cloned(),
```

Comment:
This function is correct but it is somewhat fragile and not as clean as it could be.

The `potential_commit_id` is calculated in the `commit_rule_for_certified_block()` function. This function used to be the only one that would know that the potential commit is the grandparent of the proposal. However, we are now leaking this knowledge to `construct_ledger_info()`, which is looking at the grandparent as well.

Can we have it in one place? For example, `ledger_info_for_potential_commit()` would first exercise the commit rule, and would then would either call `zero_ledger_info()` or `gen_ledger_info(block_info: &BlockInfo)`. This way, `gen_ledger_info()` would be just a mechanical thing for generating the LedgerInfo out of the given block info (that is chosen by the commit rule).
What do you think?

Reply:
I agree, this is not that awesome. I was planning on saving that for later when working on more tests / benchmarks. Rather than delay, I'll take care of it in a stacked commit in this PR.

Reply:
Let me know what you think of this update. I went somewhere in the middle for the sake of test cases.

Reply:
Looks great, thank you! (making another pass over everything again...)